services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,420 @@
+import org.eclipse.hono.service.management.tenant.Tenant;
+import org.eclipse.hono.service.management.tenant.TenantManagementService;
+import org.eclipse.hono.service.tenant.TenantService;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.AuthenticationConstants;
+import org.eclipse.hono.util.RegistrationConstants;
+import org.eclipse.hono.util.RegistryManagementConstants;
+import org.eclipse.hono.util.TenantResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import io.opentracing.Span;
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.CompositeFuture;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.mongo.FindOptions;
+import io.vertx.ext.mongo.IndexOptions;
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.UpdateOptions;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ *
+ * @see <a href="https://www.eclipse.org/hono/docs/api/tenant/">Tenant API</a>
+ * @see <a href="https://www.eclipse.org/hono/docs/api/management/">Device Registry Management API</a>
+ */
+@Component
+@Qualifier("serviceImpl")
+public final class MongoDbBasedTenantService implements TenantService, TenantManagementService, Lifecycle {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private static final int INDEX_CREATION_MAX_RETRIES = 3;
+    private MongoClient mongoClient;
+    private MongoDbCallExecutor mongoDbCallExecutor;
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an instance of the {@code MongoDbCallExecutor} and {@code MongoClient}.
+     *
+     * @param mongoDbCallExecutor An instance of the mongoDbCallExecutor.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    /**
+     * Sets the configuration properties for this service.
+     *
+     * @param config The configuration properties.
+     * @throws NullPointerException if the configuration is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final MongoDbBasedTenantsConfigProperties config) {
+        this.config = Objects.requireNonNull(config);
+    }
+
+    @Override
+    public Future<Void> start() {
+        final Promise<Void> startPromise = Promise.promise();
```

Comment:
This unused variable `startPromise` could be removed.

Reply:
done